python/ray/util/options.py

```diff
@@ -0,0 +1,202 @@
+    ),
+}
+
+
+_task_only_options = {
+    "max_calls": _counting_option("max_calls", False, default_value=0),
+    # Normal tasks may be retried on failure this many times.
+    # TODO(swang): Allow this to be set globally for an application.
+    "max_retries": _counting_option("max_retries", default_value=3),
+    # override "_common_options"
+    "num_cpus": _resource_option("num_cpus", default_value=1),
+    "num_returns": _counting_option("num_returns", False, default_value=1),
+    "object_store_memory": Option(  # override "_common_options"
+        int,
+        lambda x: x is None,
+        "Setting 'object_store_memory' is not implemented for tasks",
+        default_value=None,
+    ),
+    "retry_exceptions": Option(bool, default_value=False),
+}
+
+_actor_only_options = {
+    "concurrency_groups": Option(default_value_factory=list),
+    "lifetime": Option(str, options_only=True),
+    "max_concurrency": _counting_option("max_concurrency", False, options_only=True),
+    "max_restarts": _counting_option("max_restarts", default_value=0),
+    "max_task_retries": _counting_option("max_task_retries", default_value=0),
+    "max_pending_calls": _counting_option(
+        "max_pending_calls", options_only=True, default_value=-1
+    ),
+    "namespace": Option(str, options_only=True, default_value=None),
+}
+
+# Priority is important here because during dictionary update, same key with higher
+# priority overrides the same key with lower priority. We make use of priority
+# to set the correct default value for tasks / actors.
+
+# priority: _common_options > _actor_only_options > _task_only_options
+valid_options = {**_task_only_options, **_actor_only_options, **_common_options}
+# priority: _task_only_options > _common_options
+task_options = {**_common_options, **_task_only_options}
+# priority: _actor_only_options > _common_options
+actor_options = {**_common_options, **_actor_only_options}
+
+remote_args_error_string = (
+    "The @ray.remote decorator must be applied either with no arguments and no "
+    "parentheses, for example '@ray.remote', or it must be applied using some of "
+    f"the arguments in the list {list(valid_options.keys())}, for example "
+    "'@ray.remote(num_returns=2, resources={\"CustomResource\": 1})'."
+)
+
+
+def validate_task_options(options: Dict[str, Any], include_options_only=True):
+    """Options check for Ray tasks.
+
+    Args:
+        options: Options for Ray tasks.
+        include_options_only: If True, the options that should be only
+            used in ".options()" are included.
+    """
+    for k, v in options.items():
+        if k not in task_options:
+            if k not in valid_options:
+                raise ValueError(
+                    f"Invalid option keyword: '{k}'. {remote_args_error_string}"
+                )
+            else:
+                raise ValueError(
+                    f"The keyword '{k}' is not allowed for remote functions."
+                )
```

Comment:
I feel we should just have the same error message for both cases: Invalid option keyword {k} for remote functions, Valid ones are {task_options}

Reply:
Hmmm, I think this is the original design, because users may use options of actors for remote functions. But yes we can just use a uniformed message, the original one could be more confusing.

Reply:
fixed